sabnzbd/nzbqueue.py

```diff
@@ -896,11 +901,21 @@ def stop_idle_jobs(self):
             if len(nzo.try_list) >= nr_servers:
                 # Maybe the NZF's need a reset too?
                 for nzf in nzo.files:
+                    if nzo.removed_from_queue:
```

Comment:
Maybe we should wrap this whole section in `with NZBQUEUE_LOCK ` so we don't have to check for these things?
Because also below I wonder if we won't end up in a race condition with nzf.articles?
If one is removed during looping then it will result in a crash.

Reply:
I only added those checks for performance. Initially in my test it would go through all the files and articles in the nzo, even though the quick completion test would move the nzo out of the queue quickly because less than 100% including par2 was available.

My test consisted of an nzb where all the articles were missing on all the servers, and the lowest priority server had an invalid host name. I then disabled the invalid server when all the valid servers had tried all the articles. You have to comment out `self.info = False` in this PR to do the same test.

The locking problem I don't properly understand. Intuitively I would agree, but the more I dig into it, the stranger it seems. `sabnzbd.NzbQueue.register_article()` is called inside the article loop and did not cause problems in my tests. I assume it removed articles from the list while looping.

Also the downloader thread loops through the articles in an unlocked mode in `get_articles()` even though the decoder thread regularly deletes articles from `nzf.articles` through `NzbQueue.register_article()`. That doesn't seem to cause problems either. What am I missing?

I could change it so that the check is done before looping the articles in `nzf.get_articles()` in the downloader thread, using `nzf.articles[:]` to avoid deleting from the same list as it's looping. It would reduce some of the concurrency. It shouldn't cause a noticeable performance hit in normal circumstances if the servers parameter is changed so that it only includes active servers.

Reply:
Since this is only called in case an idle job is found, do we really need to worry about performance? I guess it doesn't hurt, the `if`'s just bothered me a tiny bit 😅

From what I can see, the only place `nzf.articles` has articles removed is using `nzf.remove_article`. Which is called from `nzo.remove_article` which is locked by `NZO_LOCK`. 
https://github.com/sabnzbd/sabnzbd/blob/be7ae3d151339fd379d1774f0a86b688887df9ff/sabnzbd/nzbstuff.py#L1190-L1191
So I think we either need to wrap this is `NZO_LOCK` or use `nzf.articles[:]`.

Reply:
I removed one of the if blocks. I also realized that removing elements from the list I'm looping over creates problems because the element position is shifted. If element 0 is removed then element 1 becomes 0 and so on. The loop has already read element 0 so the original element 1 will be skipped. If trying to remove every element then this causes half of them to be skipped silently if [:] isn't used.

Regarding the lock, I was thinking about this part: https://github.com/sabnzbd/sabnzbd/blob/develop/sabnzbd/nzbstuff.py#L432

The remove_article methods are called from other threads than the downloader. They could theoretically remove an article from nzf.articles while the downloader thread is using the get_articles() code to loop over the article list. As far as I can tell no level of the get_articles system is locked so using the NZO_LOCK before removing doesn't fix it. I don't think it would print any errors or crash if it happens but it may skip one of the remaining articles in the list. Adding [:] would avoid it.

Reply:
Btw, `NzbQueue.register_article()` is also using `nzo.remove_article()`, so NZO_LOCK will be set while an article is removed.